core/src/main/java/com/linecorp/armeria/client/HttpResponseDecoder.java

```diff
@@ -199,50 +200,73 @@ public boolean isOpen() {
          */
         @Override
         public boolean tryWrite(HttpObject o) {
+            boolean wrote = false;
             switch (state) {
                 case WAIT_NON_INFORMATIONAL:
-                    // NB: It's safe to call logBuilder.startResponse() multiple times.
-                    if (ctx != null) {
-                        ctx.logBuilder().startResponse();
-                    }
-
-                    assert o instanceof HttpHeaders && !(o instanceof RequestHeaders) : o;
-
-                    if (o instanceof ResponseHeaders) {
-                        final ResponseHeaders headers = (ResponseHeaders) o;
-                        final HttpStatus status = headers.status();
-                        if (!status.isInformational()) {
-                            state = State.WAIT_DATA_OR_TRAILERS;
-                            if (ctx != null) {
-                                ctx.logBuilder().responseHeaders(headers);
-                            }
-                        }
-                    }
+                    wrote = handleWaitNonInformational(o);
                     break;
                 case WAIT_DATA_OR_TRAILERS:
-                    if (o instanceof HttpHeaders) {
-                        state = State.DONE;
-                        if (ctx != null) {
-                            ctx.logBuilder().responseTrailers((HttpHeaders) o);
-                        }
-                    } else {
-                        if (ctx != null) {
-                            ctx.logBuilder().increaseResponseLength((HttpData) o);
-                        }
-                    }
+                    wrote = handleWaitDateOrTrailers(o);
                     break;
                 case DONE:
                     ReferenceCountUtil.safeRelease(o);
-                    return false;
+                    break;
             }
-            return delegate.tryWrite(o);
+
+            return wrote;
         }
 
         @Override
         public boolean tryWrite(Supplier<? extends HttpObject> o) {
             return delegate.tryWrite(o);
         }
 
+        private boolean handleWaitNonInformational(HttpObject o) {
+            // NB: It's safe to call logBuilder.startResponse() multiple times.
+            if (ctx != null) {
+                ctx.logBuilder().startResponse();
+            }
+
+            assert o instanceof HttpHeaders && !(o instanceof RequestHeaders) : o;
+
+            if (o instanceof ResponseHeaders) {
+                final ResponseHeaders headers = (ResponseHeaders) o;
+                final HttpStatus status = headers.status();
+                if (!status.isInformational()) {
+                    state = State.WAIT_DATA_OR_TRAILERS;
+                    if (ctx != null) {
```

Comment:
It seems unnecessary because of pre-checked condition. https://github.com/line/armeria/pull/2832/files#diff-833e9abb4a1607339097e0e3aab8b26dR226

Reply:
I don't think so because we didn't bail out when `ctx == null`.

Reply:
Oops sorry I misread. Please ignore this.

Reply:
No worries. I appreciate that you review our work! 🙇